From: Roman Yepishev <rye@gitlab.gnome.org>
Subject: [PATCH] plugins/power: Do not unidle for inactive session

NVIDIA drivers change the virtual terminal before suspend and restore
it upon resume. During this transition, the session is temporarily
inactive. The power plugin would ignore this state change, retaining
"sleep" status. When the screensaver's WakeUpScreen signal arrived
during this window, it would store "sleep" as the previous idle mode
and trigger suspension 15 seconds later.

This prevents setting up future idle actions when the session is not
active, breaking the re-suspend chain.

Closes: https://gitlab.gnome.org/GNOME/gnome-settings-daemon/-/issues/903
See: https://gitlab.gnome.org/GNOME/gnome-settings-daemon/-/merge_requests/462
---
 plugins/power/gsd-power-manager.c | 41 +++++++++++++++++++++++--------
 1 file changed, 31 insertions(+), 10 deletions(-)

diff --git a/plugins/power/gsd-power-manager.c b/plugins/power/gsd-power-manager.c
index f0718a7f93..6c053b4492 100644
--- a/plugins/power/gsd-power-manager.c
+++ b/plugins/power/gsd-power-manager.c
@@ -2279,6 +2279,35 @@ set_temporary_unidle_on_ac (GsdPowerManager *manager,
         }
 }

+static gboolean
+should_set_temporary_unidle_on_ac (GsdPowerManager *manager)
+{
+        /* Do not unidle if lid is closed */
+        if (manager->lid_is_closed)
+                return FALSE;
+
+        /* Do not unidle if we are not in an active session */
+        if (! manager->session_is_active)
+                return FALSE;
+
+        /* Unidle is already running, so we will need to reset the timer */
+        if (manager->temporary_unidle_on_ac_id != 0)
+                return TRUE;
+
+        /* Do not unidle if our current state isn't dim or blank */
+        if (manager->current_idle_mode != GSD_POWER_IDLE_MODE_BLANK &&
+            manager->current_idle_mode != GSD_POWER_IDLE_MODE_DIM)
+                return FALSE;
+
+        return TRUE;
+}
+
+static void
+update_temporary_unidle_on_ac (GsdPowerManager *manager)
+{
+        set_temporary_unidle_on_ac(manager, should_set_temporary_unidle_on_ac (manager));
+}
+
 static void
 up_client_on_battery_cb (UpClient *client,
                          GParamSpec *pspec,
@@ -2303,13 +2332,7 @@ up_client_on_battery_cb (UpClient *client,

         idle_configure (manager);

-        if (manager->lid_is_closed)
-                return;
-
-        if (manager->current_idle_mode == GSD_POWER_IDLE_MODE_BLANK ||
-            manager->current_idle_mode == GSD_POWER_IDLE_MODE_DIM ||
-            manager->temporary_unidle_on_ac_id != 0)
-                set_temporary_unidle_on_ac (manager, TRUE);
+        update_temporary_unidle_on_ac (manager);
 }

 static void
@@ -2372,7 +2395,7 @@ handle_screensaver_active (GsdPowerManager *manager,
 static void
 handle_wake_up_screen (GsdPowerManager *manager)
 {
-        set_temporary_unidle_on_ac (manager, TRUE);
+        update_temporary_unidle_on_ac (manager);
 }

 static void
--
2.47.0
